File: app/src/main/java/com/lumiyaviewer/lumiya/slproto/messages/HealthMessage.java
package com.lumiyaviewer.lumiya.slproto.messages;

import com.lumiyaviewer.lumiya.slproto.SLMessage;
import java.nio.ByteBuffer;

/**
 * Health Message
 * Tells viewer what agent health is
 *
 * <p>Template: {@code HealthMessage Low 138 Trusted Zerocoded}
 * (recovered/reference/message_template.msg).
 * <p>Viewer reference: {@code process_health_message()} in indra/newview/llviewermessage.cpp
 * (secondlife/viewer @ c179f76c01).
 */
public class HealthMessage extends SLMessage {
    public HealthData HealthData_Field;

    /** Block HealthData, Single. */
    public static class HealthData {
        public float Health; // F32
    }

    public HealthMessage() {
        this.zeroCoded = true;
        this.HealthData_Field = new HealthData();
    }

    @Override
    public int CalcPayloadSize() {
        return 8;
    }

    @Override
    public void Handle(SLMessageHandler messageHandler) {
        messageHandler.HandleHealthMessage(this);
    }

    @Override
    public void PackPayload(ByteBuffer byteBuffer) {
        // Message number: Low 138 (HealthMessage).
        byteBuffer.putShort((short) 0xFFFF);
        byteBuffer.put((byte) 0x00);
        byteBuffer.put((byte) 0x8A);
        packFloat(byteBuffer, this.HealthData_Field.Health);
    }

    @Override
    public void UnpackPayload(ByteBuffer byteBuffer) {
        this.HealthData_Field.Health = unpackFloat(byteBuffer);
    }
}

File: app/src/main/java/com/lumiyaviewer/lumiya/slproto/messages/HealthMessage.kt
package com.lumiyaviewer.lumiya.slproto.messages

import com.lumiyaviewer.lumiya.slproto.SLMessage
import java.nio.ByteBuffer

/**
 * Health Message
 * Tells viewer what agent health is
 *
 * <p>Template: {@code HealthMessage Low 138 Trusted Zerocoded}
 * (recovered/reference/message_template.msg).
 * <p>Viewer reference: {@code process_health_message()} in indra/newview/llviewermessage.cpp
 * (secondlife/viewer @ c179f76c01).
 */
open class HealthMessage : SLMessage() {
    @JvmField var HealthData_Field: HealthData = HealthData()

    /** Block HealthData, Single. */
    open class HealthData {
        @JvmField var Health: Float = 0f
    }

    init {
        zeroCoded = true
    }

    override fun CalcPayloadSize(): Int {
        return 8
    }

    override fun Handle(messageHandler: SLMessageHandler) {
        messageHandler.HandleHealthMessage(this)
    }

    override fun PackPayload(byteBuffer: ByteBuffer) {
        // Message number: Low 138 (HealthMessage).
        byteBuffer.putShort(0xFFFF.toShort())
        byteBuffer.put((0x00).toByte())
        byteBuffer.put((0x8A).toByte())
        packFloat(byteBuffer, HealthData_Field.Health)
    }

    override fun UnpackPayload(byteBuffer: ByteBuffer) {
        HealthData_Field.Health = unpackFloat(byteBuffer)
    }
}
